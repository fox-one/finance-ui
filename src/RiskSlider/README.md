# RiskSlider

## Example

```vue
<template>
  <risk-slider
    v-model="percent"
    :tips="tips"
    :scale="scale"
    ref="slider"
  />
</template>

<script lang="ts">
  import { Component, Vue } from 'vue-property-decorator';

  @Component
  export class App extends Vue {
    percent = 30;
    tips = {
      tip: 'Collateralization Ration reach',
      highlight: `${(30.21 / 50 * 100).toFixed(2)}%, itâ€™s in low-risk now.`
    };
    scale = {
      low: 0.65,
      mid: 0.2,
      high: 0.15
    };
  }
</script>
```
