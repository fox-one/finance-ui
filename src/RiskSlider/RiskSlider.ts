import {
  Component,
  Vue,
  Prop,
  Model
} from 'vue-property-decorator';
import classnames from '@utils/classnames';
import 'vue-slider-component/theme/default.css';
/* import types */
import type { CreateElement, VNode } from 'vue';

// avoid typescript error
const VueSlider = require('vue-slider-component');

@Component
export class RiskSlider extends Vue {
  @Model('change', { type: [String, Number], default: '' }) public readonly value!: string | number;

  @Prop({ type: String, default: '' }) private className!: string;
  @Prop({ type: Boolean, default: true }) private disabled!: boolean;
  @Prop({
    type: Object, default: () => ({
      low: 0.7,
      mid: 0.15,
      high: 0.15
    })
  }) private scale!: {
    low: number;
    mid: number;
    high: number;
  }
  @Prop({ type: Object, default: () => ({}) }) private tips!: {
    tip?: string;
    highlight?: string;
  }

  private get isDark() {
    return this.$vuetify?.theme?.dark;
  }

  private get colors() {
    return {
      dark: {
        low: '#56D471',
        mid: '#F79F4D',
        high: '#F67070'
      },
      light: {
        low: '#2CC94E',
        mid: '#F58721',
        high: '#F44C4C'
      }
    };
  }

  public get currentScale() {
    const val = +this.value / 100;
    if (val <= this.scale.low) {
      return 'low';
    } else if (val <= (this.scale.low + this.scale.mid)) {
      return 'mid';
    } else {
      return 'high';
    }
  }

  public render(h: CreateElement): VNode {
    const content = this.$slots.default;
    const classes = classnames('risk-slider');
    const color = this.colors[this.isDark ? 'dark' : 'light'];
    return h(
      'div',
      {
        staticClass: classes(),
        class: this.className
      },
      [
        h(
          VueSlider,
          {
            props: {
              value: this.value,
              height: 8,
              min: 0,
              max: 100,
              interval: 0.01,
              lazy: true,
              tooltip: 'none',
              disabled: this.disabled,
              railStyle: {
                'background': `linear-gradient(to right, ${color.low} 0% ${this.scale.low * 100}%, ${color.mid} ${this.scale.low * 100}% ${(this.scale.low + this.scale.mid) * 100}%, ${color.high} ${(this.scale.low + this.scale.mid) * 100}% 100%)`
              }
            }
          }
        ),
        content ?? h(
          'div',
          {
            staticClass: classes('tips', 'f-greyscale-3 f-caption')
          },
          [
            h(
              'span',
              this.tips.tip
            ),
            h(
              'span',
              {
                staticClass: classes('tips-highlight', 'font-weight-bold'),
                style: {
                  color: color[this.currentScale]
                }
              },
              this.tips.highlight
            )
          ]
        )
      ]
    );
  }
}

export default RiskSlider;