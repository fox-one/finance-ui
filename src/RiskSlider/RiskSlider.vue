<template>
  <div :class="classes()">
    <vue-slider
      v-model="_value"
      :height="8"
      :min="0"
      :max="100"
      :interval="0.01"
      :lazy="true"
      tooltip="none"
      :contained="true"
      :disabled="disabled"
      :rail-style="{
        'background': `linear-gradient(to right, ${color().low} 0% ${getScaleRatio('low') * 100}%, ${color().mid} ${getScaleRatio('low') * 100}% ${(getScaleRatio('low') + getScaleRatio('mid')) * 100}%, ${color().high} ${(getScaleRatio('low') + getScaleRatio('mid')) * 100}% 100%)`
      }"
    />
    <slot name="default">
      <div :class="classes('tips', 'f-greyscale-3 f-caption')">
        <span>{{ tips.tip }}</span>
        <span :class="classes('tips-highlight', 'font-weight-bold')" :style="{ color: color[currentScale] }">{{ tips.highlight }}</span>
      </div>
    </slot>
  </div>
</template>

<script lang="ts">
import {
  defineComponent,
  PropType,
  onMounted
} from '@vue/composition-api';
import classnames from '@utils/classnames';
import VueSlider from 'vue-slider-component';
import 'vue-slider-component/theme/default.css';

export type SCALE = {
  low: number;
  mid: number;
  high: number;
};

export type TIPS = {
  tip?: string;
  highlight?: string;
}

export default defineComponent({
  name: 'RiskSlider',
  components: {
    VueSlider
  },
  model: {
    prop: 'value',
    event: 'change'
  },
  props: {
    value: {
      type: [String, Number],
      default: ''
    },
    disabled: {
      type: Boolean,
      default: false
    },
    scale: {
      type: Object as PropType<SCALE>,
      default: () => ({
        low: 0.7,
        mid: 0.15,
        high: 0.15
      })
    },
    tips: {
      type: Object as PropType<TIPS>,
      default: () => ({})
    },
  },
  setup() {
    const classes = classnames('risk-slider');
    const colors = {
        dark: {
          low: '#56D471',
          mid: '#F79F4D',
          high: '#F67070'
        },
        light: {
          low: '#2CC94E',
          mid: '#F58721',
          high: '#F44C4C'
        }
      };
    onMounted(() => {
      console.info('RiskSlider mounted!');
    });

    return { classes, colors };
  },
  computed: {
    _value: {
      get (): string | number {
        return this.value;
      },
      set (val) {
        this.$emit('change', val);
      }
    },
    isDark(): boolean {
      return this.$vuetify?.theme?.dark;
    },
    currentScale() {
      const val = +this.value / 100;
      if (val <= this.getScaleRatio('low')) {
        return 'low';
      } else if (val <= (this.getScaleRatio('low') + this.getScaleRatio('mid'))) {
        return 'mid';
      } else {
        return 'high';
      }
    }
  },
  methods: {
    color() {
      return this.colors[this.isDark ? 'dark' : 'light']; 
    },
    getScaleRatio(scale: keyof SCALE) {
      const { low, mid, high } = this.scale;
      const total = low + mid + high;
      return this.scale[scale] / total;
    }
  }
});
</script>