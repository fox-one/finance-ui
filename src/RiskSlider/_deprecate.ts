import {
  Component,
  Vue,
  Prop,
  Model
} from 'vue-property-decorator';
import classnames from '@utils/classnames';
import 'vue-slider-component/theme/default.css';
/* import types */
import type { CreateElement, VNode } from 'vue';

// avoid typescript error
const VueSlider = require('vue-slider-component');

export type SCALE = 'low' | 'mid' | 'high';

@Component
export class RiskSlider extends Vue {
  @Model('change', { type: [String, Number], default: '' }) public readonly value!: string | number;

  @Prop({ type: String, default: '' }) private className!: string;
  @Prop({ type: Boolean, default: true }) private disabled!: boolean;
  @Prop({
    type: Object, default: () => ({
      low: 0.7,
      mid: 0.15,
      high: 0.15
    })
  }) private scale!: {
    low: number;
    mid: number;
    high: number;
  }
  @Prop({ type: Object, default: () => ({}) }) private tips!: {
    tip?: string;
    highlight?: string;
  }

  private get isDark() {
    return this.$vuetify?.theme?.dark;
  }

  private get colors() {
    return {
      dark: {
        low: '#56D471',
        mid: '#F79F4D',
        high: '#F67070'
      },
      light: {
        low: '#2CC94E',
        mid: '#F58721',
        high: '#F44C4C'
      }
    };
  }

  public get currentScale() {
    const val = +this.value / 100;
    if (val <= this.getScaleRatio('low')) {
      return 'low';
    } else if (val <= (this.getScaleRatio('low') + this.getScaleRatio('mid'))) {
      return 'mid';
    } else {
      return 'high';
    }
  }

  public getScaleRatio (scale: SCALE) {
    const { low, mid, high } = this.scale;
    const total = low + mid + high;
    return this.scale[scale] / total;
  }

  public render(h: CreateElement): VNode {
    const content = this.$slots.default;
    const classes = classnames('risk-slider');
    const color = this.colors[this.isDark ? 'dark' : 'light'];
    return h(
      'div',
      {
        staticClass: classes(),
        class: this.className
      },
      [
        h(
          VueSlider,
          {
            props: {
              value: this.value,
              height: 8,
              min: 0,
              max: 100,
              interval: 0.01,
              lazy: true,
              tooltip: 'none',
              disabled: this.disabled,
              railStyle: {
                'background': `linear-gradient(to right, ${color.low} 0% ${this.getScaleRatio('low') * 100}%, ${color.mid} ${this.getScaleRatio('low') * 100}% ${(this.getScaleRatio('low') + this.getScaleRatio('mid')) * 100}%, ${color.high} ${(this.getScaleRatio('low') + this.getScaleRatio('mid')) * 100}% 100%)`
              }
            }
          }
        ),
        content ?? h(
          'div',
          {
            staticClass: classes('tips', 'f-greyscale-3 f-caption')
          },
          [
            h(
              'span',
              this.tips.tip
            ),
            h(
              'span',
              {
                staticClass: classes('tips-highlight', 'font-weight-bold'),
                style: {
                  color: color[this.currentScale]
                }
              },
              this.tips.highlight
            )
          ]
        )
      ]
    );
  }
}

export default RiskSlider;