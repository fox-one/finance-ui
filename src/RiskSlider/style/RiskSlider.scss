@import '../../styles/variable.scss';

.#{$prefix}-risk-slider {
  width: 100%;
  &-tips {
    margin-top: 6px;
    &-highlight {
      margin-left: 2px;
    }
  }

  .vue-slider-disabled {
    opacity: 1;
  }
  .vue-slider-rail .vue-slider-process {
    background-color: transparent;
  }
  .vue-slider-rail .vue-slider-dot {
    width: 6px !important;
    height: 6px !important;
    background-color: #fff;
    border-radius: 50%;
    box-shadow: none;
  }

  .vue-slider-dot .vue-slider-dot-handle {
    width: auto;
    height: auto;
  }

  .vue-slider-dot-handle.vue-slider-dot-handle-focus {
    box-shadow: none;
  }

  .vue-slider-dot.vue-slider-dot-disabled {
    opacity: 1;
  }

  .vue-slider-dot-handle-disabled {
    background: #fff;
  }
}

.theme--dark {
  .#{$prefix}-risk-slider {
    .vue-slider-disabled {
      opacity: 1;
    }
    .vue-slider-rail .vue-slider-process {
      background-color: transparent;
    }
    .vue-slider-rail .vue-slider-dot {
      width: 6px !important;
      height: 6px !important;
      background-color: #fff;
      border-radius: 50%;
      box-shadow: none;
    }

    .vue-slider-dot-handle.vue-slider-dot-handle-focus {
      box-shadow: none;
    }

    .vue-slider-dot.vue-slider-dot-disabled {
      opacity: 1;
    }

    .vue-slider-dot-handle-disabled {
      background: #fff;
    }
  }
}