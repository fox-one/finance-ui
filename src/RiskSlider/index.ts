import RiskSlider from './RiskSlider';

export { RiskSlider } from './RiskSlider';

export default RiskSlider;
