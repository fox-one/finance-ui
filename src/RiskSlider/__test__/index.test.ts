import Vue from 'vue';
import VueCompositionAPI from '@vue/composition-api';
import { shallowMount, mount } from '@vue/test-utils';
import RiskSlider from '../';

Vue.use(VueCompositionAPI);

describe('RiskSlider', () => {
  it('renders correctly', () => {
    const wrapper = shallowMount(RiskSlider);
    expect(wrapper).toMatchSnapshot();
  });
});
