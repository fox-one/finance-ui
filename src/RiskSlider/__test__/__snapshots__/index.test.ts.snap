// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RiskSlider renders correctly 1`] = `
VueWrapper {
  "_emitted": Object {},
  "_emittedByOrder": Array [],
  "isFunctionalComponent": undefined,
}
`;
